### PR TITLE
Add Keypoint Matcher pipeline

### PR DESCRIPTION
**Pull request by sbucaille**
# What does this PR do?

Implements `keypoint-matcher` pipeline.

Quite basic for now, let me know if I should add things.
I added tests to have single and multiple pairs as well as checking it correctly fails when there is only one image provided.
Committed on top of #39968 but will be rebased on main once the fix is merged

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@qubvel @stevhliu

### CODE REVIEW

**Review by stevhliu — COMMENTED** [review]
Super nice! On the docs side, would you also like to update SuperPoint, SuperGlue, and LightGlue with the new pipeline?

**Comment by sbucaille** [comment]
@stevhliu I've added docs to SuperGlue, LightGlue and EfficientLoFTR (SuperPoint is a keypoint detector, not matcher, so it does not fall in the scope)

File: docs/source/en/model_doc/efficientloftr.md [hunk 1]
```diff
@@ -27,9 +27,26 @@ rendered properly in your Markdown viewer.
...
 
 <hfoptions id="usage">
+<hfoption id="Pipeline">
+
+```py
+import torch
+from transformers import pipeline
+
+pipeline = pipeline(
+    task="image-classification",
```
Comment (qubvel) [inline docs/source/en/model_doc/efficientloftr.md]:
```suggestion
    task="keypoint-matching",
```

File: docs/source/en/model_doc/lightglue.md [hunk 2]
```diff
@@ -29,9 +29,25 @@ You can find all the original LightGlue checkpoints under the [ETH-CVG](https://
...
 
 <hfoptions id="usage">
+<hfoption id="Pipeline">
+
+```py
+import torch
+from transformers import pipeline
+
+pipeline = pipeline(
+    task="image-classification",
```
Comment (qubvel) [inline docs/source/en/model_doc/lightglue.md]:
```suggestion
    task="keypoint-matching",
```

File: docs/source/en/model_doc/superglue.md [hunk 3]
```diff
@@ -29,9 +29,27 @@ You can find all the original SuperGlue checkpoints under the [Magic Leap Commun
...
 
 <hfoptions id="usage">
+<hfoption id="Pipeline">
+
+```py
+import torch
+from transformers import pipeline
+
+pipeline = pipeline(
+    task="image-classification",
```
Comment (qubvel) [inline docs/source/en/model_doc/superglue.md]:
```suggestion
    task="keypoint-matching",
```

File: tests/pipelines/test_pipelines_keypoint_matching.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                "http://images.cocodataset.org/val2017/000000039769.jpg",
+            ]
+        )
+
+        self.assertEqual(
+            outputs,
+            [
+                {"score": ANY(float), "label": ANY(str)},
+                {"score": ANY(float), "label": ANY(str)},
+            ],
```
Comment (qubvel) [inline tests/pipelines/test_pipelines_keypoint_matching.py]: It's not the correct format for an image-matching pipeline, is it?

**Review by qubvel — COMMENTED** [review]
Very nice, thanks for the addition of the pipeline 🔥 Great work, let's also enable pipeline tests for superglue, lightglue, efficientloftr

**Comment by sbucaille** [comment]
@qubvel didn't mean to do this review thing 😅 anyway can you enlightned me on this question ?

File: tests/pipelines/test_pipelines_keypoint_matching.py (hunk 4 above)
Comment (qubvel) [reply tests/pipelines/test_pipelines_keypoint_matching.py]: @sbucaille this is image-classification pipeline schema, and you need to update it to keypoint matching output example + replace real numbers with `type` checks

Comment (qubvel) [reply tests/pipelines/test_pipelines_keypoint_matching.py]: e.g. 

```{"score": 0.9, "label": "a dog"} -> {"score": ANY(float), "label": ANY(str)}```

**Comment by sbucaille** [comment]
@qubvel addressed the `run_pipeline_test` comment

**Review by qubvel — APPROVED** [review]

**Comment by qubvel** [comment]
cc @Rocketknight1  for review

File: src/transformers/pipelines/keypoint_matching.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        outputs = super().__call__(formatted_inputs, threshold=threshold, **kwargs)
+        if len(formatted_inputs) == 1:
+            return outputs[0]
+        return outputs
+
+    def preprocess(self, images, timeout=None):
+        images = [load_image(image, timeout=timeout) for image in images]
+        model_inputs = self.image_processor(images=images, return_tensors=self.framework)
+        if self.framework == "pt":
+            model_inputs = model_inputs.to(self.torch_dtype)
```
Comment (Rocketknight1) [inline src/transformers/pipelines/keypoint_matching.py]: In general, we've deprecated support for TF and Flax at this point. I think for new code you should just assume the framework is Torch, and raise an error if it's not, maybe in `__init__()`?

**Review by Rocketknight1 — APPROVED** [review]
PR looks good to me and tests seem clean! I made one comment but it's a nit, so feel free to ignore it.

File: src/transformers/pipelines/keypoint_matching.py (hunk 5 above)
Comment (qubvel) [reply src/transformers/pipelines/keypoint_matching.py]: addressed in https://github.com/huggingface/transformers/pull/39970/commits/cd052ed7e3c1c85f8954d3a19f3c5ace65104308
